src/System.IO.Pipes/tests/AnonymousPipeTests/AnonymousPipeTest.CrossProcess.cs

```diff
@@ -30,23 +31,78 @@ public void PingPong()
+                    for (int i = 0; i < 10; i++)
+                    {
+                        int b = inbound.ReadByte();
+                        outbound.WriteByte((byte)b);
+                    }
+                }
+                return SuccessExitCode;
+            }
         }
 
-        private static int PingPong_OtherProcess(string inHandle, string outHandle)
+        [Fact]
+        public void ServerClosesPipe_ClientReceivesEof()
         {
-            // Create the clients associated with the supplied handles
-            using (var inbound = new AnonymousPipeClientStream(PipeDirection.In, inHandle))
-            using (var outbound = new AnonymousPipeClientStream(PipeDirection.Out, outHandle))
+            using (var pipe = new AnonymousPipeServerStream(PipeDirection.Out, HandleInheritability.Inheritable))
+            using (var remote = RemoteInvoke(new Func<string, int>(ChildFunc), pipe.GetClientHandleAsString()))
             {
-                // Repeatedly read then write a byte from and to the server
-                for (int i = 0; i < 10; i++)
+                pipe.DisposeLocalCopyOfClientHandle();
+                pipe.Write(new byte[] { 1, 2, 3, 4, 5 }, 0, 5);
+
+                pipe.Dispose();
+
+                Assert.True(remote.Process.WaitForExit(30_000));
+            }
+
+            int ChildFunc(string clientHandle)
+            {
+                using (var pipe = new AnonymousPipeClientStream(PipeDirection.In, clientHandle))
                 {
-                    int b = inbound.ReadByte();
-                    outbound.WriteByte((byte)b);
+                    for (int i = 1; i <= 5; i++)
+                    {
+                        Assert.Equal(i, pipe.ReadByte());
+                    }
+                    Assert.Equal(-1, pipe.ReadByte());
                 }
+                return SuccessExitCode;
             }
-            return SuccessExitCode;
         }
 
+        [Fact]
+        public void ClientClosesPipe_ServerReceivesEof()
+        {
+            using (var pipe = new AnonymousPipeServerStream(PipeDirection.In, HandleInheritability.Inheritable))
+            using (var remote = RemoteInvoke(new Func<string, int>(ChildFunc), pipe.GetClientHandleAsString(), new RemoteInvokeOptions { CheckExitCode = false }))
+            {
+                pipe.DisposeLocalCopyOfClientHandle();
+
+                for (int i = 1; i <= 5; i++)
+                {
+                    Assert.Equal(i, pipe.ReadByte());
+                }
+                Assert.Equal(-1, pipe.ReadByte());
+
+                remote.Process.Kill();
+            }
+
+            int ChildFunc(string clientHandle)
+            {
+                using (var pipe = new AnonymousPipeClientStream(PipeDirection.Out, clientHandle))
+                {
+                    pipe.Write(new byte[] { 1, 2, 3, 4, 5 }, 0, 5);
+                }
+                Thread.CurrentThread.Join();
+                return SuccessExitCode;
```

Comment:
why return success if CheckExitcode=false.

Reply:
To return something; doesn't matter what's returned, we'll never get here, as the previous line will hang forever.

Reply:
Just an idea but it would be nice to throw UnreachableCodeBranchReachedException (or some other better name) in such situations to clearly express that fact.